lib/resolve.nix

```diff
@@ -131,133 +130,247 @@ in rec {
         depFeat = elemAt m 2;
       in if elem feat prev then
         prev
-      else if defs' ? ${feat} then
-        foldl' go ([ feat ] ++ prev) defs'.${feat}
-      else if hasPrefix "dep:" feat then
-        [ { dep = substring 4 (-1) feat; } ] ++ prev
-      else if m == null then
-        [ feat ] ++ prev
-      else if isWeak then
-        [ { dep = depName; feat = depFeat; } ] ++ prev
-      else
-        [ { dep = depName; } { dep = depName; feat = depFeat; } ] ++ prev;
+         else if defs' ? ${feat} then
+           foldl' go ([ feat ] ++ prev) defs'.${feat}
+         else if hasPrefix "dep:" feat then
+           [ { dep = substring 4 (-1) feat; } ] ++ prev
+         else if m == null then
+           [ feat ] ++ prev
+         else if isWeak then
+           [ { dep = depName; feat = depFeat; } ] ++ prev
+         else
+           [ { dep = depName; } { dep = depName; feat = depFeat; } ] ++ prev;
     fixed = mapAttrs (feat: _: go [ ] feat) defs';
   in
     fixed;
 
-  # Enable `features` in `prev` and do recursive update according to `defs`.
-  # Optional dependencies must be included in `defs`.
-  enableFeatures = pkgId: defs: prev: features:
-    foldl' (prev: feat: let
-      m = match "(.*)/.*" feat;
-      mDep = elemAt m 0;
-      nexts =
-        if m == null then
-          # Must be defined.
-          defs.${feat} or (throw ''
-            Feature '${feat}' is invalid for ${pkgId}.
-            Possible features: ${concatStringsSep "," (attrNames defs)}
-          '')
-        else
-          # Dependent features implies optional dependency to be enabled.
-          # But non-optional dependency doesn't have coresponding feature flag.
-          optional (defs ? ${mDep}) mDep;
+  parse-feature = feat: let
+    is-optional-dependency = match "dep:(.*)" feat;
+    is-dependency-feature = match "([^?]*)([?])?/(.*)" feat;
+  in
+    if is-optional-dependency != null then
+      { type= "enable-dep"; dep-name= (builtins.head is-optional-dependency);}
+    else if is-dependency-feature != null then
+      { type = "dep-feature";
+        dep-name = (elemAt is-dependency-feature 0);
+        weak = (elemAt is-dependency-feature 1) != null;
+        feat-name = (elemAt is-dependency-feature 2);
+      } else
+        { type = "normal";
+          feat-name = feat;
+        };
+
+  # Merges different feature sets
+  # by concatenating the features and or'ing the enable field.
+  # Eg:
+  # mergeChanges [
+  # { "normal" = { "a" = { "features" = [ "foo" ]; enabled = false }; };
+  #   "build" = { "b" = { "features" = []; enabled = true; }; };
+  # }
+  # { "normal" = { "a" = { "features" = [ "bar" ]; enabled = true; }; }; }
+  # ]
+  # ->
+  # { "normal" = { "a" = { "features" = [ "foo" "bar" ]; enabled = true }; };
+  #   "build" = { "b" = { "features" = []; enabled = true; }; };
+  # }
+  # TODO: this function is horribly inneficient. rewrite it using better
+  #       functions for performance
```

Comment:
No I do not have any benchmarks, because I'm not sure how I would benchmark an specific function in nix. The best I could do is benchmark the whole build time and see if it is utterly garbage, but I've used it personally for quite a while now and haven't had any problems related to performance (though the crates that I use are not that small). Do you have any suggestions about doing that?

Reply:
> how I would benchmark an specific function

We could measure the drv instantiation speed using `nix-instantiate '(mkRustPackageOrWorkspace { .. }).dev.name.bin'`. `nix-instantiate` should be used instead of flake `nix eval` to bypass the evaluation cache.